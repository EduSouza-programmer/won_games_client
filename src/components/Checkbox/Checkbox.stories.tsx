import { Story, Meta } from "@storybook/react/types-6-0";
import Checkbox, { CheckboxProps } from ".";

export default {
  title: "Checkbox",
  component: Checkbox,
  argTypes: {
    labelColor: {
      options: ["white", "black"],
      control: { type: "radio" },
    },
    label: {
      defaultValue: "TRY IT",
    },
    labelFor: {
      defaultValue: "TRY IT",
    },
    onCheck: { action: "checked" },
    value: {
      control: false,
    },
  },
  parameters: {
    viewport: {
      defaultViewport: "mobile1",
    },
    layout: "fullscreen",

    backgrounds: {
      default: "won-dark",
    },
  },
} as Meta<CheckboxProps>;

export const Default: Story<CheckboxProps> = (args) => (
  <div
    style={{
      width: "100%",
      minHeight: "100vh",
      display: "flex",
      flexDirection: "column",
      justifyContent: "center",
      alignItems: "center",
    }}
  >
    <div
      style={{
        display: "flex",
        flexDirection: "column",
        justifyContent: "space-between",
        width: "50%",
        height: "10rem",
      }}
    >
      <Checkbox label="Action" labelFor="action" isChecked />

      <Checkbox label="Adventure" labelFor="adventure" />

      <Checkbox label="Strategy" labelFor="strategy" />

      <Checkbox {...args} />
    </div>
  </div>
);
