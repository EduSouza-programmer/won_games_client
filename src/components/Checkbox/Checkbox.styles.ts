import styled, { css } from "styled-components";

import { CheckboxProps } from ".";

export const Wrapper = styled.div`
  display: flex;
  align-items: center;
`;

export const Input = styled.input`
  ${({ theme }) => css`
    display: flex;
    justify-content: center;
    align-items: center;
    appearance: none;
    width: 1.8rem;
    height: 1.8rem;
    border: 0.2rem solid ${theme.palette.darkGray.main};
    border-radius: 0.2rem;
    outline: none;
    cursor: pointer;
    position: relative;

    transition: background 0.3s ease-in-out;
    transition: border 0.2s ease-in-out;

    &::before {
      content: "";
      width: 0.6rem;
      height: 0.9rem;
      border: 0.2rem solid ${theme.palette.white.main};
      border-top: 0;
      border-left: 0;
      transform: rotate(45deg);
      position: absolute;
      top: 0.1rem;
      opacity: 0;

      transition: all 0.1s ease-in-out;
    }

    &:checked {
      border-color: ${theme.palette.primary.main};
      background: ${theme.palette.primary.main};
      &:before {
        opacity: 1;
      }
    }

    &:focus {
      box-shadow: 0 0 0.5rem ${theme.palette.primary.main};
    }
  `}
`;

export const Label = styled.label<Pick<CheckboxProps, "labelColor">>`
  ${({ theme, labelColor }) => css`
    color: ${theme.palette[labelColor!].main};
    cursor: pointer;
    margin-left: ${theme.spacings.xxsmall};
    line-height: 1;
  `}
`;
