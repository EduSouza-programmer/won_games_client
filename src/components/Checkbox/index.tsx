import { useState, useCallback, InputHTMLAttributes } from "react";
import * as S from "./Checkbox.styles";

export type CheckboxProps = {
  label?: string;
  labelFor?: string;
  labelColor?: "black" | "white";
  onCheck?: (status: boolean) => void;
  isChecked?: boolean;
  value?: string | ReadonlyArray<string> | number;
} & InputHTMLAttributes<HTMLInputElement>;

function Checkbox({
  label,
  labelFor = "",
  labelColor = "white",
  onCheck,
  isChecked = false,
  value,
  ...props
}: CheckboxProps) {
  const [checked, setChecked] = useState(isChecked);

  const onChange = useCallback(() => {
    const toogle = !checked;
    setChecked(toogle);

    if (onCheck) {
      onCheck(toogle);
    }
  }, [onCheck, checked]);

  return (
    <S.Wrapper>
      <S.Input
        onChange={onChange}
        id={labelFor}
        type="checkbox"
        checked={checked}
        value={value}
        {...props}
      />
      {!!label && (
        <S.Label labelColor={labelColor} htmlFor={labelFor}>
          {label}
        </S.Label>
      )}
    </S.Wrapper>
  );
}

export default Checkbox;
