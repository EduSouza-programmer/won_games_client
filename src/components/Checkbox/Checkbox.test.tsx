import { screen, waitFor } from "@testing-library/react";
import { renderWithTheme } from "@/utils/tests/helpers";
import userEvent from "@testing-library/user-event";

import Checkbox from ".";

describe("<Checkbox />", () => {
  it("should render with label", () => {
    const { container } = renderWithTheme(
      <Checkbox label="checkbox label" labelFor="check" />
    );

    expect(screen.getByRole("checkbox")).toBeInTheDocument();
    expect(screen.getByLabelText(/checkbox label/i)).toBeInTheDocument();
    expect(screen.getByText(/checkbox label/i)).toHaveAttribute("for", "check");

    expect(container.firstChild).toMatchSnapshot();
  });

  it("should render without label", () => {
    renderWithTheme(<Checkbox />);

    expect(screen.getByRole("checkbox").parentElement?.childElementCount).toBe(
      1
    );
  });

  it("should render a label color white by default", () => {
    renderWithTheme(<Checkbox label="checkbox label" labelFor="check" />);

    expect(screen.getByText(/checkbox label/i)).toHaveStyle({
      color: "#FAFAFA",
    });
  });
  it("should render with black label", () => {
    renderWithTheme(
      <Checkbox labelColor="black" label="checkbox label" labelFor="check" />
    );

    expect(screen.getByText(/checkbox label/i)).toHaveStyle({
      color: "#030517",
    });
  });

  it("should dispatch onCheck when status changes", async () => {
    const onCheck = jest.fn();

    renderWithTheme(<Checkbox label="checkbox" onCheck={onCheck} />);

    expect(onCheck).not.toHaveBeenCalled();

    userEvent.click(screen.getByRole("checkbox"));

    await waitFor(() => {
      expect(onCheck).toHaveBeenCalledTimes(1);
    });
    expect(onCheck).toHaveBeenCalledWith(true);
  });

  it("should dispatch onCheck called with false when isChecked passed", async () => {
    const onCheck = jest.fn();

    renderWithTheme(<Checkbox label="checkbox" onCheck={onCheck} checked />);

    userEvent.click(screen.getByRole("checkbox"));

    await waitFor(() => {
      expect(onCheck).toHaveBeenCalledTimes(1);
    });
    expect(onCheck).toHaveBeenCalledWith(false);
  });

  it("should be accessible with tab", () => {
    renderWithTheme(<Checkbox label="Checkbox" labelFor="check" />);

    expect(document.body).toHaveFocus();

    userEvent.tab();
    expect(document.body).not.toHaveFocus();
    expect(screen.getByRole("checkbox")).toHaveFocus();
  });
});
