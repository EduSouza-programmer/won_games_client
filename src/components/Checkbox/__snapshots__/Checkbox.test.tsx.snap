// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Checkbox /> should render with label 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  width: 1.8rem;
  height: 1.8rem;
  border: 0.2rem solid #2E2F42;
  border-radius: 0.2rem;
  outline: none;
  cursor: pointer;
  position: relative;
  -webkit-transition: background 0.3s ease-in-out;
  transition: background 0.3s ease-in-out;
  -webkit-transition: border 0.2s ease-in-out;
  transition: border 0.2s ease-in-out;
}

.c1::before {
  content: "";
  width: 0.6rem;
  height: 0.9rem;
  border: 0.2rem solid #FAFAFA;
  border-top: 0;
  border-left: 0;
  -webkit-transform: rotate(45deg);
  -ms-transform: rotate(45deg);
  transform: rotate(45deg);
  position: absolute;
  top: 0.1rem;
  opacity: 0;
  -webkit-transition: all 0.1s ease-in-out;
  transition: all 0.1s ease-in-out;
}

.c1:checked {
  border-color: #F231A5;
  background: #F231A5;
}

.c1:checked:before {
  opacity: 1;
}

.c1:focus {
  box-shadow: 0 0 0.5rem #F231A5;
}

.c2 {
  color: #FAFAFA;
  cursor: pointer;
  margin-left: 0.8rem;
  line-height: 1;
}

<div
  class="c0"
>
  <input
    class="c1"
    id="check"
    type="checkbox"
    value=""
  />
  <label
    class="c2"
    for="check"
  >
    checkbox label
  </label>
</div>
`;
